domain/leaseservice.go

```diff
@@ -0,0 +1,116 @@
+	if err := ctx.Err(); err != nil {
+		return ctx.Err()
+	}
+
+	// The leaseCtx will be cancelled when the lease is no longer held by the
+	// lease holder. This may or may not be the same as the holderName for the
+	// lease. That check is done by the Token checker.
+	leaseCtx, leaseCancel := context.WithCancel(ctx)
+	defer leaseCancel()
+
+	leaseChecker := s.leaseChecker()
+
+	// Check if the lease is held by the holder, if not, return immediately.
+	token := leaseChecker.Token(leaseName, holderName)
+	if err := token.Check(); err != nil {
+		return err
+	}
+
+	// Start will be closed when we start waiting for the lease to expire.
+	// If the lease is not held, the function will return immediately and
+	// the context will be cancelled.
+	start := make(chan struct{})
+
+	// WaitUntilExpired will be run against the leaseName, we've already
+	// checked that the lease is held by the holder, but there is a race
+	// between the prior check and the start of the lease waiting. To ensure
+	// that after we've waited that we still hold the lease, we need to check
+	// that the lease is still held by the holder. Then we can guarantee that
+	// the lease is held by the holder for the duration of the function.
+	// Although convoluted this is necessary to ensure that the lease is held
+	// by the holder for the duration of the function.
+	// The context will be cancelled when the lease is no longer held by
+	// the lease holder for the lease name.
+
+	waitCtx, waitCancel := context.WithCancel(ctx)
+	defer waitCancel()
+
+	waitErr := make(chan error)
+	go func() {
+		// This guards against the case that the lease has changed state
+		// before we run the function.
+		err := leaseChecker.WaitUntilExpired(waitCtx, leaseName, start)
+
+		// Ensure that the lease context is cancelled when the wait has
+		// completed. We do this as quick as possible to ensure that the
+		// function is cancelled as soon as possible.
+		leaseCancel()
+
+		// The waitErr might not be read, so we need to provide another way
+		// to collapse the goroutine. Using the waitCtx this goroutine will
+		// be cancelled when the function is complete.
+		select {
+		case <-waitCtx.Done():
+			return
+		case waitErr <- err:
+		}
+	}()
+
+	select {
+	case <-leaseCtx.Done():
+		return leaseCtx.Err()
+	case err := <-waitErr:
+		if err == nil {
+			// This shouldn't happen, but if it does, we need to return an
+			// error. If we're attempting to wait whilst holding the lease,
+			// before running the function and then wait return nil, we don't
+			// know if the lease is held by the holder or what state we're in.
+			return fmt.Errorf("unable to wait for lease to expire whilst holding lease")
+		}
+		return err
+	case <-start:
+	}
+
+	// Ensure that the lease is held by the holder before proceeding.
+	// We're guaranteed that the lease is held by the holder, otherwise the
+	// context will have been cancelled.
+	if err := token.Check(); err != nil {
```

Comment:
Can't we do the check just once here? I think it would be safe having started the logic above. That ensures nothing is _changing_, all we have to do check that the condition is now still valid. It's one DB hit instead of two.

Reply:
> Can't we do the check just once here?

You can remove the first check not the second. There is a small race between the first token check against the wait call. If the leaseholder changes, you won't know that and you'll be acting on incorrect information. So the correct way to check this is, wait and then ensure the holder is correct. That way you can guarantee that it's the same one (the context will be canceled because the lease has changed).

The first check is purely to prevent the wait call, so you'll be either running an upfront token call or wait + token call in the worst case. The happy path case would be token + wait + token or wait + token. I don't mind which, we just need to pick the poison.

Reply:
You had it right - remove the first check; keep the second. So wait + token.